consensus/src/messages/handlers.rs

```diff
@@ -304,144 +304,190 @@ impl<N: Network> Handle<N, ResponseChunk, Arc<RwLock<Blockchain>>> for RequestCh
+        if block_number < election_head {
             log::info!(
-                current_head,
-                requested_block_number = self.block_number,
-                "Requested txn proof from the future",
+                block_number = block_number,
+                "Requested txn proof that corresponds to a finalized epoch, should use the election block instead",
             );
             return ResponseTransactionsProof {
                 proof: None,
                 block: None,
             };
         }
-
-        let block;
-
-        if Policy::is_election_block_at(self.block_number) {
-            log::info!(
-                election_block_number = self.block_number,
-                len = hashes.len(),
-                "Requested txn proof from finalized epoch",
-            );
-            // If we were provided the block number of an election block it has to be already finalized
-            block = blockchain
-                .chain_store
-                .get_block_at(self.block_number, false, None)
-                .ok();
-
-            if block.is_none() {
-                log::info!("Could not find the desired block to create the txn proof");
-                return ResponseTransactionsProof {
-                    proof: None,
-                    block: None,
-                };
-            }
-            verifier_state = None;
-        } else if Policy::is_macro_block_at(self.block_number) {
+        block_number
+    } else {
+        log::info!(
+            block_number = block_number,
+            len = hashes.len(),
+            "Requested txn proof from current batch",
+        );
+        // If we were provided a block number corresponding to a micro block,
+        // it needs to correspond to at least the previous batch (we allow the
+        // previous batch also, to not fail when the client is a few blocks behind
+        // and we are already at or over a macro block).
+        // If the requested block is older than the previous batch, the requester
+        // should use the latest checkpoint block number instead.
+        if block_number < macro_head - Policy::blocks_per_batch() {
             log::info!(
-                checkpoint_block_number = self.block_number,
-                len = hashes.len(),
-                "Requested txn proof from finalized checkpoint block",
+                block_number = block_number,
+                "Requested txn proof from finalized batch, should use a checkpoint block instead",
             );
+            return ResponseTransactionsProof {
+                proof: None,
+                block: None,
+            };
+        }
+        block_number
+    };
+
+    let block = blockchain
+        .chain_store
+        .get_block_at(proving_block_number, false, None)
+        .ok();
+
+    if let Some(ref block) = block {
+        // We have some extra work for a macro block
+        if Policy::is_macro_block_at(proving_block_number) {
```

Comment:
Does this not also return `true` for election blocks, where we don't need it?

```suggestion
        if Policy::is_macro_block_at(proving_block_number) && !Policy::is_election_block(proving_block_number) {
```

Reply:
True! good catch